### ISSUE TITLE
Clippy reports a while_let_on_iterator with example code that does not compile

### ISSUE THREAD

**Issue by Kixiron** [rust-lang/rust-clippy#5540]
### Bug

Clippy reports a `while_let_on_iterator` with example code that does not compile. [Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ec5c98be3ff0df3e1967bd766e70205b)

#### Original code

```rust
let mut slice = [1, 2, 3, 4, 5, 6, 7, 8].windows(2);
while let Some([a, b]) = slice.next() {
    println!("{}, {}", a, b);
}
```

#### Clippy Error

```rust
warning: this loop could be written as a `for` loop
 --> src/main.rs:4:30
  |
4 |     while let Some([a, b]) = slice.next() {
  |                              ^^^^^^^^^^^^ help: try: `for [a, b] in slice { .. }`
  |
  = note: `#[warn(clippy::while_let_on_iterator)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#while_let_on_iterator
```

#### Code with clippy solution

```rust
let mut slice = [1, 2, 3, 4, 5, 6, 7, 8].windows(2);
for [a, b] in slice {
    println!("{}, {}", a, b);
}
```

#### Compile Error

```rust
error[E0005]: refutable pattern in `for` loop binding: `&[]`, `&[_]` and `&[_, _, _, ..]` not covered
  --> src/main.rs:10:9
   |
10 |     for [a, b] in slice {
   |         ^^^^^^ patterns `&[]`, `&[_]` and `&[_, _, _, ..]` not covered
```

### Clippy Version

```sh
❯ cargo clippy -V
clippy 0.0.212 (6651c1b9 2020-04-15)
```

**Comment by alex-700** [comment]
[is_refutable](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/utils/mod.rs#L934) looks buggy. For example, it thinks that a slice pattern is refutable only when it contains a refutable inner pattern. I'll deal with it by the weekends.

**Comment by flip1995** [comment]
Duplicate of #3780

@alex-700 There was already a fix attempt in #3784, but was sadly abandoned.
